2018/socrates-linz/README.md

```diff
@@ -0,0 +1,49 @@
+# Implementation
+
+- Show "main.go" our implementation
+
+# Unit Testing
+
+- Unit testing is usually you get X in and check that Y comes out.
+- Show "main_test.go our unit tests
+- Run `go test -v -coverprofile=coverage.out` -> coverage: 58.3% of statements -> WHAT?!
+- Run `go tool cover -html=coverage.out`
+- Problem: You are just testing what you think is correct (e.g. only things that are defined in the specification).
+
+# Mutation-testing
+
+- https://en.wikipedia.org/wiki/Mutation_testing -> `Mutation testing (or mutation analysis or program mutation) is used to design new software tests and evaluate the quality of existing software tests. Mutation testing involves modifying a program in small ways. Each mutated version is called a mutant and tests detect and reject mutants by causing the behavior of the original version to differ from the mutant. This is called killing the mutant. Test suites are measured by the percentage of mutants that they kill.`
+- https://github.com/zimmski/go-mutesting my implementation for Go
+- Run `go-mutesting --debug github.com/symflower/sessions/2018/socrates-linz/csvler` -> `The mutation score is 0.333333 (3 passed, 6 failed, 3 duplicated, 0 skipped, total is 9)` -> There are just 3 "mutations" that are tested by our unit-tests but 6 cases are not! (3 are duplicates -> go-mutesting did some work that lead to the same program/coverage/results)
+
+# Fuzzing
+
+- https://en.wikipedia.org/wiki/Fuzzing -> `Fuzzing or fuzz testing is an automated software testing technique that involves providing invalid, unexpected, or random data as inputs to a computer program.`
+- My implementation of a fuzzing/model-based testing and delta-debugging tool in one binary: https://github.com/zimmski/tavor The file format is defined here https://github.com/zimmski/tavor/blob/master/doc/format.md (there is also an API).
+- Show "example.tavor"
+- Run `tavor --format-file example.tavor fuzz`
+- Run `tavor --format-file example.tavor fuzz --strategy AllPermutations`
+- Show "csvler.tavor"
+- Do the csvler.tavor with just two columns -> not likely to run into the bug
+- Run `tavor --format-file csvler.tavor --max-repeat 10 fuzz` -> generates randomly CSV files with our format. And repeats things at a maximum 10 times.
```

Comment:
Maybe "random" is correct rather than randomly, but this depends on whether the word references generate or CSV files...